### PR TITLE
Phase utils

### PR DESCRIPTION
**Pull request by luojing1211**
This is the Pull request for the phase utilities.

### CODE REVIEW

**Comment by mhvk** [comment]
For a test file, one can do:
```
pint_utils = pytest.importorskip('scintillometry.utils.pint_utils')
```
(This presumes `pint_utils` will try to import `pint` and raise an `ImportError`. If that is not the case, just `pytest.importorskip('pint')`)

And in `.travis.yml` one can add a test in which `PINT` is a dependency.

**Comment by luojing1211** [comment]
@mhvk This failure of travis-ci test is interesting. it did not find numpy when installing the PINT package. Maybe I don't quite understand how travis-ci sets up the environment. Should I install numpy in the "before-install" section?

**Comment by mhvk** [comment]
Yes, I think in `before_install` nothing really is there yet. Possibly the easiest is to adjust `MAIN_CMD`, something like
```
MAIN_CMD='./install_pint.sh; python setup.py'
```

File: scintillometry/utils/phase_utils.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+import warnings
+
+import numpy as np
+import astropy.units as u
+import astropy.time as time
+
+try:
+    from .pint_utils import PintUtils
+except ModuleNotFoundError:
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: I think this should be `except ImportError` to get rid of the failure.

Alternatively, and perhaps better, you could do the import inside `PintPhase.__init__`, so that it doesn't even try unless the user actually wants to use PINT. In that case, of course you do not need to check for the import error at all.

**Comment by mhvk** [comment]
@luojing1211 - great to hear that pint now works with astropy 3.1. Here, see my in-line comment about what I think is causing the test failure. (Aside: rather busy with teaching now, but will try to actually review the meat of this!)

File: install-pint.sh [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
```
Comment (mhvk) [inline install-pint.sh]: This file is not needed any more

File: scintillometry/integration.py [hunk 3]
```diff
@@ -406,6 +407,9 @@ def __init__(self, ih, n_phase, phase, average=True,
 
     def _read_frame(self, frame_index):
         # Read frame in phased directly, bypassing its ``read`` method.
+        print('Stack: time {}, sample {}/{}'
```
Comment (mhvk) [inline scintillometry/integration.py]: Debugging statement?

File: scintillometry/tests/test_phaseutils.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        self.polyco_pu = PolycoPhase(self.polyco_file)
+
+
+class TestPhase(TestPhaseBase):
+    def test_one_phase(self):
+        # Compute PINT phase and f0 for one time
+        pint_phase_one = self.pint_pu(self.times[0])
+        pint_f0_one = self.pint_pu.apparent_spin_freq(self.times[0])
+        assert isinstance(pint_phase_one, u.Quantity), \
+            "PINT phase did not reture the expected list."
```
Comment (mhvk) [inline scintillometry/tests/test_phaseutils.py]: Here and below: "reture" -> "return"

File: scintillometry/utils/phase_utils.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+       Note
+       ----
+       This method provides high precision phase calculation. The fractional
+       phase is between -0.5 to 0.5.
+    """
+    def __init__(self, par_file, obs, obs_freq, ephem='de436',
+                 bipm_version='BIPM2015'):
+        self.par_file = par_file
+        self.pu = PintUtils(self.par_file)
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: I'm usually against local imports, but here I think it makes a lot of sense to just import `PintUtils` in here (and thus omit the import above)._

File: scintillometry/utils/pint_utils.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+                               'tdb_method': tdb_method}
+        self.control_params.update(kwargs)
+        # Load toa file
+        self.tim_file = tim_file
+
+    @property
+    def par_file(self):
+        return self._par_file
+
+    @par_file.setter
```
Comment (mhvk) [inline scintillometry/utils/pint_utils.py]: I guess you have the setter to validate the parameters (at least, I'm not sure why we would allow after-the-fact setting), but I find it confusing that setting `par_file` ends up producing a `model` attribute. My sense would be to just do this in `__init__`.

File: scintillometry/utils/pint_utils.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        if val is not None:
+            self.model = model.get_model(val)
+        else:
+            self.model = None
+
+    @property
+    def tim_file(self):
+        return self._tim_file
+
+    @tim_file.setter
```
Comment (mhvk) [inline scintillometry/utils/pint_utils.py]: As above for the par file.

File: scintillometry/utils/pint_utils.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+           Raises
+           ------
+           RunTimeError
+               If the timestamps are not provided.
+           ValueError
+               If the observatory name or observing frequency are not provided,
+               when using timestamps.
+        """
+        # This enables user input control parameters from the this method.
```
Comment (mhvk) [inline scintillometry/utils/pint_utils.py]: This comment is not very useful - why is it necessary to override? I guess in part I'm not sure why there is a separate `PintUtils` and `PintPhase` class. I think `PintUtils` is meant to be the equivalent of the my `Predictor` class?

File: scintillometry/utils/pint_utils.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+        if control_params != {}:
+            self.control_params.update(control_params)
+
+        # Load toas from tim file.
```
Comment (mhvk) [inline scintillometry/utils/pint_utils.py]: I see what this does, but am not sure I understand why, in the context of the current scintillometry routines, it is useful!

File: scintillometry/integration.py (hunk 3 above)
Comment (luojing1211) [reply scintillometry/integration.py]: That is a debugging statement. I will remove it.

File: scintillometry/utils/pint_utils.py (hunk 8 above)
Comment (luojing1211) [reply scintillometry/utils/pint_utils.py]: The `PintUtils` are designed to provide some convenient functions for basic PINT usages (e.g., compute the apparent spin frequency, compute the time delays, etc). If in the scintillometry analysis, we only care about the pulse phase, then the PintUtils are not necessary.

File: scintillometry/utils/pint_utils.py (hunk 6 above)
Comment (luojing1211) [reply scintillometry/utils/pint_utils.py]: Maybe I was thinking from the point of taking advantage of PINT when I am writing this part. Sometimes, we want to use a new model for the same set of data. Reading a lot of TOAs is actually time-consuming.  That is why the after-the-fact setting is enabled.

Comment (mhvk) [reply scintillometry/utils/pint_utils.py]: Yes, I can see that point. But I think we then need to think a bit about how to make the interface similar to `Predictor` (may well be easiest to adapt predictor, not at all sold on it!).

Comment (luojing1211) [reply scintillometry/utils/pint_utils.py]: I think I can change the architecture as follows:
```
pint_modle = PintModel(parfile, control_param=control_parameters)
pint_modle includes all the functions (phase, delay,...)
phase = pint_model.phase(times, or tim_files)
```
I think this makes `PintUtils` kind of like `Predictor`

File: scintillometry/integration.py [hunk 10]
```diff
@@ -5,6 +5,7 @@
 import warnings
 
 import numpy as np
+from astropy.time import Time
```
Comment (mhvk) [inline scintillometry/integration.py]: Why is this import here? I don't think `Time` is used directly in `integration.py`

File: scintillometry/utils/phase_utils.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+
+class PintPhase(object):
+    """A utility class designated for computing phase using PINT.
+       Parameter
+       ---------
+       par_file : str
+           TEMPO/TEMPO2 style parameter fileself.
+       obs : str
+           Observatory name or observatory codeself.
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: extraneous `self` at end

File: scintillometry/utils/phase_utils.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+       obs_freq : float or `~astropy.units.Quantity`.
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: Just call this `frequency`?

File: scintillometry/utils/phase_utils.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+           Observing frequency default units is 'MHz'
+       ephem : str
+           The solar system ephemeris version.
+       bipm_version : str
+           The BIPM clock correction version.
+
+       Note
+       ----
+       This method provides high precision phase calculation. The fractional
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: High precision as compared to what? I think it is important here to tell that it returns two-part phases.

File: scintillometry/utils/phase_utils.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+       phase is between -0.5 to 0.5.
+    """
+    def __init__(self, par_file, obs, obs_freq, ephem='de436', include_bipm=True,
+                 bipm_version='BIPM2015', include_gps=True, planets=False,
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: I think all arguments after `bipm_version` should be removed - they can be given as `kwargs` anyway.

File: scintillometry/utils/phase_utils.py [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+       ---------
+       par_file : str
+           TEMPO/TEMPO2 style parameter fileself.
+       obs : str
+           Observatory name or observatory codeself.
+       obs_freq : float or `~astropy.units.Quantity`.
+           Observing frequency default units is 'MHz'
+       ephem : str
+           The solar system ephemeris version.
+       bipm_version : str
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: I think examples here are useful. Alternatively, remove this (and `ephem` too?) and rely on defaults that can be overridden with `kwargs` (and tell where to look up what those are).

File: scintillometry/utils/phase_utils.py [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+        self.obs = obs
+        self.obs_freq = obs_freq
+
+    def __call__(self, t):
+        """Compute the apparent phase at one or more time stemp
+
+           Parameter
+           ---------
+           t : `astropy.time.Time` object
+               The input time stemps.
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: stemps -> stamps

(Same several more times.)

File: scintillometry/utils/phase_utils.py [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+                                  **kwargs)
+        self.obs = obs
+        self.obs_freq = obs_freq
+
+    def __call__(self, t):
+        """Compute the apparent phase at one or more time stemp
+
+           Parameter
+           ---------
+           t : `astropy.time.Time` object
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: object -> instance (or remove altogether)

File: scintillometry/utils/phase_utils.py [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+               The input time stemps.
+
+           Return
+           ------
+           The apparent pulse phase at time t. The phases are return as
+           'integer phase, fractional phase'
+        """
+        toas = self._make_toas(t)
+        ph = self.model.phase(toas)
+        return u.Quantity((ph.int, ph.frac), u.cycle)
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: I think it actually makes more sense to return two quantities, which has the advantage that a copy can be avoided, i.e.,
```
return u.Quantity(ph.int, u.cycle, copy=False), u.Quantity(ph.frac, u.cycle, copy=False)
```

File: scintillometry/utils/phase_utils.py [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+            ---------
+            t : `astropy.time.Time` object
+                The input time stemps.
+
+            Return
+            ------
+            The apparent spin frequency at time t.
+        """
+        toas = self._make_toas(t)
+        apprnt_f0 = self.model.d_phase_d_toa(toas)
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: I'd spell out the name or just do `return self.model....`

File: scintillometry/utils/phase_utils.py [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+               The input time stemps.
+
+           Return
+           ------
+           The apparent pulse phase at time t. The phases are return as
+           'integer phase, fractional phase'.
+        """
+        t = t._apply(np.atleast_1d)
+        ph = self.polyco(t)
+        with u.set_enabled_equivalencies([(u.cycle, None)]):
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: This is clever, but perhaps also here return two separate parts:
```
ph_int, ph_frac = divmod(ph, 1)
return u.Quantity(ph_int, u.cycle, copy=False), u.Quantity(ph_frac, u.cycle, copy=False)
```

File: scintillometry/utils/pint_utils.py (hunk 9 above)
Comment (mhvk) [reply scintillometry/utils/pint_utils.py]: We now always pass in the output of `make_toa_list`, so let's only accept that for now.

Actually, I'm not sure why we keep that as a separate function rather than just put it here?

File: scintillometry/utils/phase_utils.py [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+           ---------
+           t : `astropy.time.Time` object
+               The input time stemps.
+
+           Return
+           ------
+           The apparent pulse phase at time t. The phases are return as
+           'integer phase, fractional phase'.
+        """
+        t = t._apply(np.atleast_1d)
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: Is the 1-D time needed for `Polyco`?

Comment (luojing1211) [reply scintillometry/utils/phase_utils.py]: One may want to see the phase only at one time. This line just makes sure the input time is iteratable.

File: scintillometry/utils/phase_utils.py (hunk 14 above)
Comment (luojing1211) [reply scintillometry/utils/phase_utils.py]: I think the user should see it. Because the older version BIPM files do not cover the recent date.  I feel it is kind of dangerous to hide it. Maybe it is the time to push Astropy time object to handle it?

File: scintillometry/utils/phase_utils.py (hunk 11 above)
Comment (luojing1211) [reply scintillometry/utils/phase_utils.py]: Sorry, my editor has a bad habit to guess what I want to do.

File: scintillometry/utils/phase_utils.py (hunk 21 above)
Comment (mhvk) [reply scintillometry/utils/phase_utils.py]: OK. Though I think `Polyco` should handle that natively... Anyway, can be done in another PR.

File: scintillometry/utils/phase_utils.py (hunk 14 above)
Comment (mhvk) [reply scintillometry/utils/phase_utils.py]: OK, so let's just remove the argument beyond it, i.e., the ones not currently documented.

If it isn't there already, could you raise an issue over at astropy that describes what the BIPM things do?

**Comment by luojing1211** [comment]
@mhvk , is this pull request ready. I think it is time to merge it.

File: install-pint.sh (hunk 2 above)
Comment (mhvk) [reply install-pint.sh]: This file still needs to be removed.

File: scintillometry/tests/data/B1937+21_58245.dat [hunk 22]
```diff
@@ -0,0 +1,432 @@
+B1937+21    5-May-18  220000.00   58243.91666666660            70.966367 -0.714 -6.252
```
Comment (mhvk) [inline scintillometry/tests/data/B1937+21_58245.dat]: Could we reduce this file in size by just keeping four entries or so?

File: scintillometry/tests/data/B1937+21_NANOGrav_11yv1.gls.par [hunk 23]
```diff
@@ -0,0 +1,1059 @@
+PSR              B1937+21
```
Comment (mhvk) [inline scintillometry/tests/data/B1937+21_NANOGrav_11yv1.gls.par]: Why do we need two par files? Can we just stick with the much shorter NGC-* one?

File: scintillometry/tests/data/B1937_polyco.dat [hunk 24]
```diff
@@ -0,0 +1,336 @@
+B1937+21    6-MAY-18   12736.00   58244.06083333330            71.090617 -0.723 -5.224
```
Comment (mhvk) [inline scintillometry/tests/data/B1937_polyco.dat]: Why do we need 2 polyco example files?

File: scintillometry/tests/data/J0613-sim.tim [hunk 25]
```diff
@@ -0,0 +1,1464 @@
+FORMAT 1
```
Comment (mhvk) [inline scintillometry/tests/data/J0613-sim.tim]: Do we need the file at all? If so, again just keep 3 lines or so.

File: scintillometry/tests/data/NGC6440E.tim [hunk 26]
```diff
@@ -0,0 +1,64 @@
+MODE 1
```
Comment (mhvk) [inline scintillometry/tests/data/NGC6440E.tim]: I guess this one is different from the other `.tim` file, but is this really needed for `scintillometry`? If so, just a few lines should suffice.

File: scintillometry/tests/test_phaseutils.py (hunk 4 above)
Comment (mhvk) [reply scintillometry/tests/test_phaseutils.py]: This comment still needs addressing.

File: scintillometry/tests/test_phaseutils.py [hunk 27]
```diff
@@ -0,0 +1,82 @@
+# Licensed under the GPLv3 - see LICENSE
+"""Full-package tests of pint_utils sources."""
```
Comment (mhvk) [inline scintillometry/tests/test_phaseutils.py]: docstring seems copied from `test_pint_utils`

File: scintillometry/tests/test_pintutils.py [hunk 28]
```diff
@@ -0,0 +1,70 @@
...
+        self.start_time = Time('2018-05-06T23:00:00', format='isot',
+                               scale='utc')
+        self.times = self.start_time + np.arange(1000) * 1.50 * u.min
+        self.obs = 'AO'
+        self.freq = 1400
+
+    def test_make_toalist(self):
+        toalist = pint_utils.make_toa_list(self.times, self.obs, self.freq)
+        assert len(toalist) == 1000, \
+            "TOA list did not made properly"
```
Comment (mhvk) [inline scintillometry/tests/test_pintutils.py]: error message not grammatical; I actually do not think it is necessary to have the messages for obvious errors, so would suggest just to remove it.

File: scintillometry/tests/test_pintutils.py [hunk 29]
```diff
@@ -0,0 +1,70 @@
...
+
+    def test_time_input(self):
+        pt = pint_utils.PintToas()
+        toalist = pint_utils.make_toa_list(self.times, self.obs, self.freq)
+        toas = pt.make_toas(toa_list=toalist)
+        assert toas.ntoas == 1000, \
+            "TOAs did not input properly from timestamps"
+        assert list(set(toas.get_obss())) == ['arecibo'], \
```
Comment (mhvk) [inline scintillometry/tests/test_pintutils.py]: Just check the `set` directly, as in `set(toas.get_obss()) == {'arecibo'}`

File: scintillometry/tests/test_pintutils.py [hunk 30]
```diff
@@ -0,0 +1,70 @@
...
+            "Observatory did not input correctly."
+        assert list(set(toas.get_freqs())) == [1400 * u.MHz], \
```
Comment (mhvk) [inline scintillometry/tests/test_pintutils.py]: Same

File: scintillometry/tests/test_pintutils.py [hunk 31]
```diff
@@ -0,0 +1,70 @@
...
+            "Observing frequency did not input correctly."
+
+    def test_no_input(self):
+        # Test Exception
```
Comment (mhvk) [inline scintillometry/tests/test_pintutils.py]: This comment is not very useful: the code below obviously tests for an exception, but what is actually wrong? Let this comment tell what it tests (empty input? if so, this would normally not be a `RuntimeError`...)

File: scintillometry/tests/test_pintutils.py [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+        with pytest.raises(RuntimeError):
+            pt = pint_utils.PintToas()
```
Comment (mhvk) [inline scintillometry/tests/test_pintutils.py]: This first part is not expected to fail, correct? If so, move it outside of the `with` statement.

File: scintillometry/tests/test_pintutils.py [hunk 33]
```diff
@@ -0,0 +1,70 @@
...
+        assert toas.ntoas == 62, \
+            "TOAs did not input properly from tim file"
+
+    def test_control_param(self):
+        # Test from default control param values
+        pt = pint_utils.PintToas()
+        toalist = pint_utils.make_toa_list(self.times, self.obs, self.freq)
+        toas = pt.make_toas(toa_list=toalist)
+        assert toas.ephem == 'de436', \
+            "Fail to initialize with default control parameters."
```
Comment (mhvk) [inline scintillometry/tests/test_pintutils.py]: Fail -> Failed (here and below)

File: scintillometry/utils/phase_utils.py [hunk 34]
```diff
@@ -0,0 +1,142 @@
+# Licensed under the GPLv3 - see LICENSE
+
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: Please add a docstring!

File: scintillometry/utils/phase_utils.py [hunk 35]
```diff
@@ -0,0 +1,142 @@
...
+import warnings
+
+import numpy as np
+import astropy.units as u
+import astropy.time as time
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: Is `time` used at all? I think it is not such a good idea to import as `time`, since we use the argument `time` quite often.  I would do `from astropy.time import Time` (here if needed, everywhere below)

File: scintillometry/utils/phase_utils.py (hunk 14 above)
Comment (mhvk) [reply scintillometry/utils/phase_utils.py]: This comment still stands

File: scintillometry/utils/phase_utils.py [hunk 36]
```diff
@@ -0,0 +1,142 @@
...
+
+try:
+    from .pint_utils import PintToas, make_toa_list
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: I think these import should be in `PintPhase__init__` - this file should not bypass the import failures.

File: scintillometry/utils/phase_utils.py (hunk 21 above)
Comment (mhvk) [reply scintillometry/utils/phase_utils.py]: Are you sure there is actually a problem with 0-D times? Is it in `Polyco` or somewhere else? I don't know that anything should assume an iterable.

File: scintillometry/utils/phase_utils.py [hunk 37]
```diff
@@ -0,0 +1,142 @@
...
+            ---------
+            t : `astropy.time.Time` instance
+                The input time stamps.
+
+            Return
+            ------
+            The apparent spin frequency at time t.
+        """
+        t = t._apply(np.atleast_1d)
+        apprnt_f0 = self.polyco(t, deriv=1)
```
Comment (mhvk) [inline scintillometry/utils/phase_utils.py]: Still dislike this half-abbreviation. Call it `f` or just `return self.polyco(...).to(...)`

File: scintillometry/utils/pint_utils.py (hunk 9 above)
Comment (mhvk) [reply scintillometry/utils/pint_utils.py]: This comment still stands - why do we want to support `.tim` files?

File: scintillometry/utils/phase_utils.py (hunk 36 above)
Comment (luojing1211) [reply scintillometry/utils/phase_utils.py]: importing them in the `PintPhase.__init__` does not keep the modules after the` __init__`. In` _make_toas`, I will need to use make_toa_list. So moving `import make_toa_list` to ` _make_toas()`

**Comment by mhvk** [comment]
See https://github.com/luojing1211/scintillometry/pull/1 for a PR to your branch.

Separately, I wondered about the names: maybe this shouldn't be `utils`, but a separate `phases`? I think this can stand on its own.  It would then have a `core.py` (current `phase_utils`), everything in which is imported in `__init__`, and a `pint_toa.py` and `predictor.py`.

**Comment by mhvk** [comment]
@luojing1211 - I renamed `pint_utils` to `pint_toas` and also included `phases` in the documentation - that turned out not to work locally and after updating things it turned out there were a lot of missing links (not only for phases... see #70). Anyway, now all seems OK.

I'll merge this once tests pass...

**Comment by mhvk** [comment]
Bizarre. Works locally but not here. Seems to be something with travis not working correctly. Will investigate later.

**Comment by mhvk** [comment]
Even more bizarre - it seems it sorted itself overnight. OK, will go ahead and merge so we have this available!

Though: I do think we still need to rewrite the integration classes a little so that they can deal with two-part phase, no? Right now, one would have to add the integer and fractional parts together.

I wonder if we should have a `Phase` class that works a bit like `Time` (not so strange, the integer part are pulsar "days"), so that we can subtract phases, divide into them, etc.
